### ISSUE TITLE
[Bug] git dependencies with prepare scripts

### ISSUE THREAD

**Issue by andreialecu** [yarnpkg/berry#1477]
- [ ] I'd be willing to implement a fix

**Describe the bug**

I have a dependency to a github repository, which has a `prepare` script. Installing it works fine on `ubuntu`, but I have a dockerfile that runs `Alpine Linux` and it seems to fail to build on that platform, resulting in a cryptic error:

```
➤ YN0013: │ @typegoose/typegoose@https://github.com/typegoose/typegoose.git#commit:3ed06e5097ab929f69755676fee419318aaec73a can't be found in the cache and will be fetched from the remote repository
➤ YN0018: │ @typegoose/typegoose@https://github.com/typegoose/typegoose.git#commit:3ed06e5097ab929f69755676fee419318aaec73a: The remote archive doesn't match the expected checksum
```

I believe it happens due to a build failure on Alpine, but I'm unsure why that is, and yarn isn't helpful enough to show the log by default.

Adding `enableInlineBuilds: true` to `.yarnrc.yml` shows some additional context about a package related to `mongodb-memory-server` not being available on Alpine, but it seems it should be just a warning. I'm unsure what the reason for the actual build failure is, but I didn't have too much time to investigate.

**To Reproduce**

Repro at: https://github.com/andreialecu/yarn-berry-git-prepare

Unfortunately can't use sherlock because it only fails on Alpine. But you can use the Dockerfile there to see the error.

**Screenshots**

**Environment if relevant (please complete the following information):**

 - OS: Alpine Linux
 - Node version: 13
 - Yarn version: 2.0.0-rc.34

**Additional context**

Adding the `.yarn/cache` directory to the docker context is a viable workaround. In the repro it's in `.dockerignore` to ensure it reproduces.

**Comment by andreialecu** [comment]
Additionally there's an inconsistency in how `yarn.lock` is generated. I ended up with this `diff` at one point:

![image](https://user-images.githubusercontent.com/697707/84590323-35ada800-ae3e-11ea-986d-d2bcf022efc5.png)
 
I think what happened was that initially the dependency was in `package.json` and I just ran `yarn install`, but on a second run I ran `yarn add @typegoose/typegoose@https://github.com/typegoose/typegoose.git#master`.

Notice that the only diff is a change from `:` to `=`

**Comment by andreialecu** [comment]
Another issue here is that running `yarn up @typegoose/typegoose` reverts the dependency to the npm version rather than refreshing it to the latest commit, as yarn v1 used to do. Not sure if this is by design or not.

**Comment by arcanis** [comment]
> Not sure if this is by design or not.

Yep, expected (we quickly talked about this on Discord).

> Notice that the only diff is a change from `:` to `=`

This is expected, as we changed the `:` to `=` in new builds. Since you run `yarn add`, we regenerate the lockfile entry per the new requirements.

> The remote archive doesn't match the expected checksum

That means the archive we generate (using either of `npm pack` or `yarn pack`, depending on the target repository) doesn't match the expected checksum, making your installs non-deterministic. Unless the repository you're installing uses Yarn 2, there's little we can do to fix this as it really depends on the project and the package manager.

I think there isn't anything else actionable, so I'll go ahead and close the issue!

**Comment by andreialecu** [comment]
I was thinking at a minimum a better error message could be thrown. It took several hours of debugging to understand why that error was occuring.

Perhaps something like: 

```
Preparing dependency (running prepare script) for ...
Checksum of prepared dependency ... does not match ... 
Check build output for ... in yarn-error.log
```

It's also possible that a build may not be entirely deterministic. Perhaps a timestamp is embedded at build time. A way to disable the checksum check just for a specific dependency might also be a good idea to have.

**Comment by andreialecu** [comment]
For anyone that may be running into similar issues, the problem was that `npm pack` was failing to produce a proper archive on Alpine Linux because of permission problems.

The default user on Alpine seems to be root, and `yarn` detects that the `package-lock.json` file in the repo and runs `npm pack` to build it.

[According to npm documentation](https://docs.npmjs.com/misc/scripts#user):
> If npm was invoked with root privileges, then it will change the uid to the user account or uid specified by the user config, which defaults to nobody. Set the unsafe-perm flag to run scripts with root privileges.

The `nobody` user will then have no permissions to write files, thus silently failing the build. 

Basically it would log a `WARN` with 

`npm WARN cannot run in wd ...`

But **still** pack the resulting archive (which was missing compiled files), and thus result in a broken build.

The solution was to run `npm set unsafe-perm true` prior to running `yarn install` on Alpine.

**Comment by gorjuspixels** [comment]
My Dockerfile copied the project into a home directory (https://github.com/yarnpkg/berry/issues/2187). This `YN0018` error was resolved by making sure the project is using its own directory like `/home/${USER}/app`
